### ISSUE TITLE
`MailExecution` keeps running forever when used with Flow triggers, as the task was designed for listeners

### ISSUE THREAD

**Issue by npranav10** [kestra-io/plugin-notifications#101]
## TL;DR:

* I am trying setup an email notifier in case of failure of the flow whose ID is `fail_flow`.
* I have the email credentials stored as global variables just for testing purposes.
* `MailSend` plugin works but `MailExecution` does not.
* I am wondering if my method of using `MailExecution` plugin is flawed.

---

## Flow (that is built to be failed):

Here is a sample flow. Because of the command `exit 1`, the flow is always destined to be a failed execution.

``` yaml
id: fail_flow
namespace: ipynbs
tasks:
 - id: test1
   type: io.kestra.plugin.scripts.shell.Commands
   commands:
      - exit 1 # implying processing of 'fail_flow' fails
```

## MailSend

The `MailSend` plugin seems to be working fine if my flow's code is in the following manner:

```yaml
id: fail_flow
namespace: ipynbs
tasks:
 - id: test1
   type: io.kestra.plugin.scripts.shell.Commands
   commands:
      - exit 1 # implying processing of 'fail_flow' fails
errors:
  - id: send_email
    type: io.kestra.plugin.notifications.mail.MailSend
    from: "{{ globals['gmail-username'] }}"
    to: "{{ globals['gmail-username'] }}"
    username: "{{ globals['gmail-username'] }}"
    password: "{{ globals['gmail-password'] }}"
    host: smtp.gmail.com
    port: 587
    transportStrategy: SMTP_TLS
    subject: "Kestra workflow '{{flow.id}}' failed"
    htmlTextContent: "Failure alert for flow {{ flow.namespace }}.{{ flow.id }} with ID {{ execution.id }}"
```

## MailExecution

But when I try to use the `MailExecution` plugin in the folllowing manner (see yaml below), the `failure_alert` flow gets triggered but keeps running forever and I have to manually kill it:

```yaml
id: failure_alert
namespace: ipynbs
tasks:
  - id: send_alert
    type: io.kestra.plugin.notifications.mail.MailExecution
    to: "{{ globals['gmail-username'] }}"
    from: "{{ globals['gmail-username'] }}"
    subject: "The workflow execution {{trigger.executionId}} failed for the flow {{trigger.flowId}} in the namespace {{trigger.namespace}}"
    host: smtp.gmail.com
    port: 587
    transportStrategy: SMTP_TLS
    username: "{{ globals['gmail-username'] }}"
    password: "{{ globals['gmail-password'] }}"
triggers:
  - id: listen_fail_flow
    type: io.kestra.core.models.triggers.types.Flow
    conditions:
      - type: io.kestra.core.models.conditions.types.ExecutionFlowCondition
        namespace: ipynbs
        flowId: fail_flow
      - type: io.kestra.core.models.conditions.types.ExecutionStatusCondition
        in:
          - FAILED
```

![image](https://github.com/kestra-io/plugin-notifications/assets/22992723/c23344be-18c1-4aeb-9a9f-14d7ae6537d4)

Here is the log for the `failure_alert` flow:

![image](https://github.com/kestra-io/plugin-notifications/assets/22992723/177ff070-5fa2-4fcd-aac1-e187e8b46d95)

---

## Future work:

In future I am planning to replace the following condition

```yaml
      - type: io.kestra.core.models.conditions.types.ExecutionFlowCondition
        namespace: ipynbs
        flowId: fail_flow
```

with

```yaml
      - type: io.kestra.core.models.conditions.types.ExecutionNamespaceCondition
        namespace: ipynbs
```

so that I get an email on failure of any flow beloning to the `ipynbs` namespace.

**Comment by anna-geller** [comment]
thanks for the issue, so well described!

I wonder, can you try setting `sessionTimeout` to a higher value e.g. 20000? this will allow to wait a bit longer for the response from email server

**Comment by npranav10** [comment]
Hi Anna,

Looks like `sessionTimeout` isn't affecting the way the flow works.

I tried setting the value to 1000 (1 sec), 20000 (20 sec), 30000 (30 sec) - regardless of which I get the following output on the logs:

![image](https://github.com/kestra-io/plugin-notifications/assets/22992723/14378c47-ccff-41bd-90dd-2568e9555d8f)

**Comment by anna-geller** [comment]
you are spot on, this is a bug

I checked and unfortunately, the MailExecution seems to have been written in a way that works only with listeners (which we are deprecating soon), but not yet with triggers.

for now, you can validate that this would work:

```yaml
id: fail_flow
namespace: dev

tasks:
 - id: test1
   type: io.kestra.plugin.scripts.shell.Commands
   runner: PROCESS
   commands:
      - exit 1

listeners:
  - tasks:
      - id: send_execution_alert
        type: io.kestra.plugin.notifications.mail.MailExecution
        from: contact@annageller.com
        to: anna@kestra.io
        username: contact@annageller.com
        host: mail.privateemail.com
        port: 465 # 587 
        password: xxxxx
        sessionTimeout: 10000
        subject: "The workflow execution {{execution.id}} failed for the flow {{flow.id}} in the namespace {{flow.namespace}}"
    conditions:
      - type: io.kestra.core.models.conditions.types.ExecutionStatusCondition
        in:
          - FAILED
          - WARNING
```

**Comment by npranav10** [comment]
Thanks.

The listener task worked perfectly.

How do I setup a flow so that the listener checks for failed execution of all flows in a given namespace, instead of attaching the listener to each flow in a namespace ?

**Comment by anna-geller** [comment]
Listeners is a feature we plan to deprecate https://kestra.io/docs/migrations/listeners  

So the ideal solution would be if for now, you could leverage the `MailSend` along with triggers, and once we fix the bug, you can update to `MailExecution`.

**Comment by loicmathieu** [comment]
Note that all the notification *Execution tasks are designed to be used inside a listener as they have default alert messages that use the current execution information. 
So if you use MailExecution (or any other tasks that end with Execution in this plugin) in a trigger you will have the execution information of the triggered flow not the original flow.

**Comment by loicmathieu** [comment]
So I found what's going on: the MailExecution (and all other *Execution tasks in this plugin) will try to load an execution based on the configured `executionId` property (which is the current execution ID by default). 
It does this with a retry, waiting for the flow to ends during 10mn.

So if you use it as is, it will wait until it's execution to be terminated during 10 minutes ... which will never occurs as the execution can only be terminated if the task that is waiting for the execution is still running. It waits for itself to ends which will never be the case. Normally after 10 minutes you should see an exception.

So, for this task to be used in a Flow trigger, you must configure it to use the `trigger.executionId` variable and not the flow `execution.id` variable so it will use the original execution not the triggered execution.

Using this flow will make it work:

```
id: failure_alert
namespace: ipynbs
tasks:
  - id: send_alert
    type: io.kestra.plugin.notifications.mail.MailExecution
    to: "{{ globals['gmail-username'] }}"
    from: "{{ globals['gmail-username'] }}"
    subject: "The workflow execution {{trigger.executionId}} failed for the flow {{trigger.flowId}} in the namespace {{trigger.namespace}}"
    host: smtp.gmail.com
    port: 587
    transportStrategy: SMTP_TLS
    username: "{{ globals['gmail-username'] }}"
    password: "{{ globals['gmail-password'] }}"
    executionId: "{{ trigger.executionId }}"
triggers:
  - id: listen_fail_flow
    type: io.kestra.core.models.triggers.types.Flow
    conditions:
      - type: io.kestra.core.models.conditions.types.ExecutionFlowCondition
        namespace: ipynbs
        flowId: fail_flow
      - type: io.kestra.core.models.conditions.types.ExecutionStatusCondition
        in:
          - FAILED
    ```

So this task is currently designed to be used by a listener, we'll need to discuss what you should do but as we cannot change the default for now I will add documentation clarifications.
